ManageAxes.js

```diff
@@ -161,3 +171,80 @@ ManageAxes.resetAxes = async function()
+    const geomTransf3d = await WSM.Transf3d.MakeTranslationTransform(vec3d);
+    // get the inverse transform for use later
+    const inverseGeomTransf3d = await WSM.Transf3d.Invert(geomTransf3d);
+
+    // check if the transform from the geom to the world origin is null - if so, stop here
+    const isGeomTransf3dNull = await WSM.Vector3d.IsNull(vec3d)
+    if (isGeomTransf3dNull) {
+        // show a message that nothing was changed
+        await FormIt.UI.ShowNotification("The origin is already centered below this geometry.\nNo changes were made.", FormIt.NotificationType.Information, 0);
+
+        // no need to go any further
+        return;
+    }
+
+    // if we got this far, changes will be made to the model, so start a new undo state
+    await FormIt.UndoManagement.BeginState();
+
+    // get all the non-owned objects in this history
+    const objectIds = await WSM.APIGetAllNonOwnedReadOnly(editingHistoryId);
+    // and move them to the world origin
+    await WSM.APITransformObjects(editingHistoryId, objectIds, geomTransf3d);
+
+    // get all the instances of this history
+    const instancesOfHistory = await WSM.APIGetAllAggregateTransf3dsReadOnly(editingHistoryId, 0);
+
+    // for each instance of this history...
+    for (let i = 0; i < instancesOfHistory.paths.length; i++) {
+        // how many levels "deep" is this instance from history 0
+        const historyDepthIndex = instancesOfHistory.paths[i].ids.length - 1;
+        const instanceObjectHistoryId = instancesOfHistory.paths[i].ids[historyDepthIndex];
+
+        // get the instance's transform and the inverse
+        const instanceTransf3d = instancesOfHistory.transforms[i];
+        const inverseInstanceTransf3d = await WSM.Transf3d.Invert(instanceTransf3d);
+
+        // multiply the geom inverse transform, the inverse instance transform, and the instance transform
+        const newTransf3dPartial = await WSM.Transf3d.Multiply(inverseGeomTransf3d, inverseInstanceTransf3d);
+        const newTransf3dFinal = await WSM.Transf3d.Multiply(instanceTransf3d, newTransf3dPartial);
+
+        // transform the instance by the final transf3d
+        await WSM.APITransformObject(instanceObjectHistoryId.History, instanceObjectHistoryId.Object, newTransf3dFinal);
+    }
+
+    // reset the local origin (this gets moved along with the non-owned objects) 
+    // so the local origin appears at the bottom center of the geometry
+    const defaultLCS = await WSM.Geom.MakeRigidTransform(await WSM.Geom.Point3d(0, 0, 0), await WSM.Geom.Vector3d(1, 0, 0), await WSM.Geom.Vector3d(0, 1, 0), await WSM.Geom.Vector3d(0, 0, 1));
+    await WSM.APISetLocalCoordinateSystem(editingHistoryId, defaultLCS);
+
+    // hack: end group edit mode and start again to force show the updated origin position
+    await FormIt.GroupEdit.EndEditInContext();
+    await FormIt.GroupEdit.SetInContextEditingPath(editingPath);
+
+    await FormIt.UndoManagement.EndState("Manage Axes - Re-Origin");
```

Comment:
This should be moved to line 239 so it goes with the changes. It gets tricky if you change context while in a begin/end state block since you could by mistake add changes to the same undo in a new history.  (There is no bug here yet but if is set up to be a potential problem).

Reply:
Done.